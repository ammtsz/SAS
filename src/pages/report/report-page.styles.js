import styled from "styled-components";
import { CardContainer } from "../../assets/css";
import {
  navbarHeight,
  mq_xs,
} from "../../assets/css/variables";


export const PageReport = styled.section`
  display: flex;
  justify-content: center;
  padding-top: 63px;
  padding-bottom: 50px;

  min-height: calc(100vh - ${navbarHeight});

  @media (max-width: ${mq_xs}) {
    padding: 0;
  }
`;

export const CardReport = styled(CardContainer)`
  position: relative;
  display: flex;
  flex-direction: column;

  width: min(534px, 100vw);
  height: 100%;
`;
