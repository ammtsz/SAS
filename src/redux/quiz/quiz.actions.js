import { QuizActionsTypes } from "./quiz.types";

// REDUCERS
export const actionSetQuizActive = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_ACTIVE,
  payload: data,
});
export const actionSetQuizLoading = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_LOADING,
  payload: data,
});
export const actionSetQuizCategory = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_CATEGORY,
  payload: data,
});
export const actionSetQuizDifficulty = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_DIFFICULTY,
  payload: data,
});
export const actionSetQuizQuestionNumber = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_QUESTION_NUMBER,
  payload: data,
});
export const actionSetQuizPromotion = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_PROMOTION,
  payload: data,
});

export const actionSetQuizCurrentQuestion = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_CURRENT_QUESTION,
  payload: data,
});
export const actionSetQuizCurrentOptions = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_CURRENT_OPTIONS,
  payload: data,
});
export const actionSetQuizToken = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_TOKEN,
  payload: data,
});
export const actionSetQuizQuestionsDatas = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_QUESTIONS_DATAS,
  payload: data,
});
export const actionSetQuizError = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_QUIZ_ERROR,
  payload: data,
});

export const actionResetQuiz = () => ({
  type: QuizActionsTypes.REDUCER_RESET_QUIZ,
});


// SAGAS
export const actionGetNewQuestion = (data) => ({
  type: QuizActionsTypes.SAGA_GET_NEW_QUESTION,
  payload: data,
});
export const actionCheckAnswer = (data) => ({
  type: QuizActionsTypes.SAGA_CHECK_ANSWER,
  payload: data,
});
export const actionUpdateQuizReport = () => ({
  type: QuizActionsTypes.SAGA_UPDATE_QUIZ_REPORT,
});
export const actionUpdateQuizResumeReport = () => ({
  type: QuizActionsTypes.SAGA_UPDATE_QUIZ_RESUME_REPORT,
});
export const actionGoToNextQuestion = (data) => ({
  type: QuizActionsTypes.SAGA_GO_TO_NEXT_QUESTION,
  payload: data,
});
export const actionFinishQuiz = (data) => ({
  type: QuizActionsTypes.SAGA_FINISH_QUIZ,
  payload: data,
});
export const actionResumeQuiz = (data) => ({
  type: QuizActionsTypes.SAGA_RESUME_QUIZ,
  payload: data,
});
