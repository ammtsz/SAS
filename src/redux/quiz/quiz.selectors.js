import { createSelector } from "reselect";

const selectQuiz = (state) => state.quiz;

export const selectQuizActive = createSelector(
  [selectQuiz],
  (quiz) => quiz.active
);

export const selectQuizCategory = createSelector(
  [selectQuiz],
  (quiz) => quiz.category
);

export const selectQuizQuestionNumber = createSelector(
  [selectQuiz],
  (quiz) => quiz.questionNumber
);

export const selectQuizDifficulty = createSelector(
  [selectQuiz],
  (quiz) => quiz.difficulty
);

export const selectQuizPromotion = createSelector(
  [selectQuiz],
  (quiz) => quiz.promotion
);

export const selectQuizQuestionDatas = createSelector(
  [selectQuiz],
  (quiz) => quiz.questionDatas
);

export const selectQuizCurrentQuestion = createSelector(
  [selectQuiz],
  (quiz) => quiz.currentQuestion
);

export const selectQuizCurrentOptions = createSelector(
  [selectQuiz],
  (quiz) => quiz.currentOptions
);
export const selectQuizToken = createSelector(
  [selectQuiz],
  (quiz) => quiz.token
);
