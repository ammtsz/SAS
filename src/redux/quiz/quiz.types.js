export const QuizActionsTypes = {
    REDUCER_SET_ACTIVE: "REDUCER_SET_ACTIVE",
    REDUCER_SET_CATEGORY: "REDUCER_SET_CATEGORY",
    REDUCER_SET_QUESTION_NUMBER: "REDUCER_SET_QUESTION_NUMBER",
    REDUCER_SET_DIFFICULTY: "REDUCER_SET_DIFFICULTY",
    REDUCER_SET_PROMOTION: "REDUCER_SET_PROMOTION",
    REDUCER_SET_QUESTION_DATAS: "REDUCER_SET_QUESTION_DATAS",
    REDUCER_SET_CURRENT_QUESTION: "REDUCER_SET_CURRENT_QUESTION",
    REDUCER_SET_CURRENT_OPTIONS: "REDUCER_SET_CURRENT_OPTIONS",
    REDUCER_SET_TOKEN: "REDUCER_SET_TOKEN",

    REDUCER_SET_QUIZ_ERROR: "REDUCER_SET_QUIZ_ERROR",
    REDUCER_RESET_QUIZ: "REDUCER_RESET_QUIZ",

    SAGA_FETCH_NEW_QUESTION: "SAGA_FETCH_NEW_QUESTION",
    SAGA_CHECK_ANSWER: "SAGA_CHECK_ANSWER",
    SAGA_RIGHT_ANSWERS_ACTIONS: "SAGA_RIGHT_ANSWERS_ACTIONS",
    SAGA_WRONG_ANSWERS_ACTIONS: "SAGA_WRONG_ANSWERS_ACTIONS",
    SAGA_GET_CURRENT_OPTIONS: "SAGA_GET_CURRENT_OPTIONS",
    SAGA_GO_TO_NEXT_QUESTION: "SAGA_GO_TO_NEXT_QUESTION",
    SAGA_FINISH_QUIZ: "SAGA_FINISH_QUIZ",
    SAGA_RESUME_QUIZ: "SAGA_RESUME_QUIZ",
}
