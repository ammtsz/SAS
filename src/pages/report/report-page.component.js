import React from "react";
import { PageReport, CardReport } from "./report-page.styles";
import ReportHeader from "./card-header.report/card-header.report.component";
import ReportBody from "./card-body.report/card-body.report.component";

import { connect } from "react-redux";
import { actionFinishQuiz } from "../../redux/quiz/quiz.actions";

import { withRouter } from "react-router-dom";
import { createStructuredSelector } from "reselect";
import { selectReports } from "../../redux/categories/categories.selectors";
import { selectQuizCategory } from "../../redux/quiz/quiz.selectors";
import {
  selectReportReview,
  selectReportCategory,
} from "../../redux/report/report.selectors";
import { actionResetReport } from "../../redux/report/report.actions";

const mapStateToProps = createStructuredSelector({
  reports: selectReports,
  quizCategory: selectQuizCategory,
  reportReview: selectReportReview,
  reportCategory: selectReportCategory,
});

const mapDispatchToProps = (dispatch) => ({
  finishQuiz: (data) => dispatch(actionFinishQuiz(data)),
  resetReport: (data) => dispatch(actionResetReport(data)),
});

const Report = (props) => {
  const { reportReview, reportCategory, quizCategory } = props;

  let category = reportReview ? reportCategory : quizCategory;

  return (
    <PageReport>
      <CardReport>
        <ReportHeader review={reportReview} category={category}/>
        <ReportBody {...props} category={category} />
      </CardReport>
    </PageReport>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Report));
