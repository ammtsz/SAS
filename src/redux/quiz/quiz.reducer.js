import { QuizActionsTypes } from "./quiz.types";

const INITIAL_STATE = {
  active: false,
  loading: false,
  category: null,
  difficulty: "medium",
  questionNumber: 1,
  promotion: "1", //0 = got last one wrong, 1 = first question on that level, 2 = got last one right
  currentQuestion: {},
  currentOptions: [],
  questionsAnswered: [],
  error: null,
};

const quizReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case QuizActionsTypes.REDUCER_SET_ACTIVE:
      return { ...state, active: action.payload };

    case QuizActionsTypes.REDUCER_SET_LOADING:
      return { ...state, loading: action.payload };

    case QuizActionsTypes.REDUCER_SET_CATEGORY:
      return { ...state, category: action.payload };

    case QuizActionsTypes.REDUCER_SET_DIFFICULTY:
      return { ...state, difficulty: action.payload };

    case QuizActionsTypes.REDUCER_SET_QUESTION_NUMBER:
      return { ...state, questionNumber: action.payload };

    case QuizActionsTypes.REDUCER_SET_PROMOTION:
      return { ...state, promotion: action.payload };

    case QuizActionsTypes.REDUCER_SET_CURRENT_QUESTION:
      return { ...state, currentQuestion: action.payload };

    case QuizActionsTypes.REDUCER_SET_CURRENT_OPTIONS:
      return { ...state, currentOptions: action.payload };

    case QuizActionsTypes.REDUCER_SET_QUESTIONS_ANSWERED:
      return { ...state, questionsAnswered: action.payload };

    case QuizActionsTypes.REDUCER_RESET_QUIZ:
      return INITIAL_STATE;

    case QuizActionsTypes.REDUCER_SET_QUIZ_ERROR:
      return { ...state, error: action.payload };

    default:
      return state;
  }
};

export default quizReducer;
