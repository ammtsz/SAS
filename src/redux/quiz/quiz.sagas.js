import { all, call, takeLatest, put, select } from "redux-saga/effects";
import { fetchQuestion } from "../../api/trivia";
import { rsf } from "../../firebase/firebase.utils";
import {
  actionSetQuizError,
  actionSetQuizDifficulty,
  actionSetQuizPromotion,
  actionSetQuizCurrentQuestion,
  actionSetQuizCurrentOptions,
  actionSetQuizQuestionNumber,
  actionResetQuiz,
  actionSetQuizToken,
  actionSetQuizActive,
  actionSetQuizCategory,
  actionSetQuizQuestionsDatas,
  actionSetQuizLoading,
} from "./quiz.actions";
import { QuizActionsTypes } from "./quiz.types";
import {
  selectQuizCurrentQuestion,
  selectQuizDifficulty,
  selectQuizPromotion,
  selectQuizQuestionNumber,
  selectQuizCategory,
  selectQuizQuestionsDatas,
  selectQuizToken,
} from "./quiz.selectors";
import { selectReports } from "../categories/categories.selectors";
import { actionSetCategoriesReport } from "../categories/categories.actions";
import { selectUserDatas } from "../user/user.selectors";

// UTILS

export function* fetchQuestionFn(difficulty, category) {
  try {
    const token = yield select(selectQuizToken);
    const fetchedQuestion = yield fetchQuestion(difficulty, category, token);
    yield updateToken(token, fetchedQuestion.token);
    return fetchedQuestion.results;
  } catch (error) {
    yield put(actionSetQuizError(error));
    return {};
  }
}
export function* updateToken(stateToken, fetchedToken) {
  try {
    if (stateToken !== fetchedToken) {
      yield put(actionSetQuizToken(fetchedToken));
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* saveQuestionsDatas(fetchedQuestion) {
  try {
    yield put(actionSetQuizCurrentQuestion(fetchedQuestion));
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* getCurrentOptions() {
  try {
    const currentQuestion = yield select(selectQuizCurrentQuestion);
    let allOptions = [];

    allOptions.push({ id: 0, option: currentQuestion.correct_answer });
    currentQuestion.incorrect_answers.forEach((option, index) => {
      allOptions.push({ id: index + 1, option });
    });

    yield put(
      actionSetQuizCurrentOptions(
        allOptions.sort((a, b) => (a.option > b.option ? 1 : -1))
      )
    );
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}

export function* rightAnswerActions() {
  try {
    const quizPromotion = yield select(selectQuizPromotion);
    if (quizPromotion === "2") {
      yield increaseDifficulty();
      yield put(actionSetQuizPromotion("1"));
    } else {
      yield put(actionSetQuizPromotion("2"));
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* increaseDifficulty() {
  try {
    const quizDifficulty = yield select(selectQuizDifficulty);
    switch (quizDifficulty) {
      case "easy":
        yield put(actionSetQuizDifficulty("medium"));
        break;
      case "medium":
        yield put(actionSetQuizDifficulty("hard"));
        break;
      default:
        yield put(actionSetQuizDifficulty(quizDifficulty));
        break;
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* wrongAnswerActions() {
  try {
    const quizPromotion = yield select(selectQuizPromotion);
    if (quizPromotion === "0") {
      yield decreaseDifficulty();
      yield put(actionSetQuizPromotion("1"));
    } else {
      yield put(actionSetQuizPromotion("0"));
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* decreaseDifficulty() {
  const quizDifficulty = yield select(selectQuizDifficulty);
  try {
    switch (quizDifficulty) {
      case "hard":
        yield put(actionSetQuizDifficulty("medium"));
        break;
      case "medium":
        yield put(actionSetQuizDifficulty("easy"));
        break;
      default:
        yield put(actionSetQuizDifficulty(quizDifficulty));
        break;
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}

export function* getReportDatas() {
  try {
    let easy = { rights: 0, wrongs: 0 };
    let medium = { rights: 0, wrongs: 0 };
    let hard = { rights: 0, wrongs: 0 };
    let rights = 0;
    let wrongs = 0;

    const quizQuestionsDatas = yield select(selectQuizQuestionsDatas);

    quizQuestionsDatas.forEach((question) => {
      if (question.difficulty === "easy") {
        if (question.selected_answer === "0") {
          easy = { ...easy, rights: easy.rights + 1 };
          rights += 1;
        } else {
          easy = { ...easy, wrongs: easy.wrongs + 1 };
          wrongs += 1;
        }
      } else if (question.difficulty === "medium") {
        if (question.selected_answer === "0") {
          medium = { ...medium, rights: medium.rights + 1 };
          rights += 1;
        } else {
          medium = { ...medium, wrongs: medium.wrongs + 1 };
          wrongs += 1;
        }
      } else if (question.difficulty === "hard") {
        if (question.selected_answer === "0") {
          hard = { ...hard, rights: hard.rights + 1 };
          rights += 1;
        } else {
          hard = { ...hard, wrongs: hard.wrongs + 1 };
          wrongs += 1;
        }
      }
    });

    return {
      difficulty: {
        easy: { rights: easy.rights, wrongs: easy.wrongs },
        medium: { rights: medium.rights, wrongs: medium.wrongs },
        hard: { rights: hard.rights, wrongs: hard.wrongs },
      },
      rights,
      wrongs,
    };
  } catch (error) {
    yield put(actionSetQuizError(error));
    return {};
  }
}
export function* updateReportOnDB(report) {
  try {
    const userDatasState = yield select(selectUserDatas);
    if (userDatasState) {
      const userDatas = yield getUserDatasFromFirebase(userDatasState.id)
      yield call(rsf.firestore.updateDocument, `users/${userDatasState.id}`, {
        ...userDatas,
        reports: { ...userDatas.reports, ...report },
      });
    } else {
      localStorage.setItem("trivia", JSON.stringify(report));
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* getUserDatasFromFirebase(id) {
  try {
    const userSnapshot = yield call(
      rsf.firestore.getDocument,
      `users/${id}`
    );
    return userSnapshot.data();
  } catch (error) {
    yield put(actionSetQuizError(error));
    return {}
  }
}

// CALLED
export function* getNewQuestion({ payload: { difficulty, category } }) {
  try {
    yield put(actionSetQuizLoading(true));
    const fetchedQuestion = yield fetchQuestionFn(difficulty, category);
    yield saveQuestionsDatas(fetchedQuestion);
    yield getCurrentOptions(fetchedQuestion);
    yield put(actionSetQuizLoading(false));
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* checkAnswer(selectedOption) {
  try {
    if (selectedOption.payload === "0") {
      yield rightAnswerActions();
    } else {
      yield wrongAnswerActions();
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* updateQuizReport() {
  try {
    const quizQuestionNumber = yield select(selectQuizQuestionNumber);
    const difficulty = yield select(selectQuizDifficulty);
    const category = yield select(selectQuizCategory);
    const reports = yield select(selectReports);
    const questions_datas = yield select(selectQuizQuestionsDatas);
    const token = yield select(selectQuizToken);
    const promotion = yield select(selectQuizPromotion);
    const results = yield getReportDatas();

    const resume =
      quizQuestionNumber < 10
        ? {
            token,
            promotion,
            difficulty,
          }
        : null;

    const categoryReport = {
      category_name: category.name,
      questions_datas,
      results,
      resume,
    };
    const newReport = { ...reports, [category.id]: categoryReport };
    yield put(actionSetCategoriesReport(newReport));
    yield updateReportOnDB(newReport);
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* updateQuizResumeReport() {
  try {
    const quizQuestionNumber = yield select(selectQuizQuestionNumber);
    const difficulty = yield select(selectQuizDifficulty);
    const category = yield select(selectQuizCategory);
    const reports = yield select(selectReports);
    const token = yield select(selectQuizToken);
    const promotion = yield select(selectQuizPromotion);
    const lastQuestion = yield select(selectQuizCurrentQuestion);

    const resume =
      quizQuestionNumber < 10
        ? {
            token,
            promotion,
            difficulty,
            lastQuestion,
          }
        : null;

    const newReport = {
      ...reports,
      [category.id]: { ...reports[category.id], resume },
    };
    yield put(actionSetCategoriesReport(newReport));
    yield updateReportOnDB(newReport);
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* goToNextQuestion() {
  try {
    const quizQuestionNumber = yield select(selectQuizQuestionNumber);
    const difficulty = yield select(selectQuizDifficulty);
    const category = yield select(selectQuizCategory);
    yield getNewQuestion({ payload: { difficulty, category: category.id } });
    yield updateQuizResumeReport();
    yield put(actionSetQuizQuestionNumber(quizQuestionNumber + 1));
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* finishQuiz() {
  try {
    yield put(actionResetQuiz());
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* resumeQuiz(category) {
  try {
    const reports = yield select(selectReports);
    const quizDatas = reports[category.payload.id];

    yield put(actionSetQuizActive(true));
    yield put(actionSetQuizCategory(category.payload));
    yield put(actionSetQuizDifficulty(quizDatas.resume.difficulty));
    yield put(actionSetQuizPromotion(quizDatas.resume.promotion));
    yield put(
      actionSetQuizQuestionNumber(quizDatas.questions_datas.length + 1)
    );
    yield put(actionSetQuizToken(quizDatas.resume.token));
    yield put(actionSetQuizQuestionsDatas(quizDatas.questions_datas));

    if (quizDatas.resume.lastQuestion) {
      yield put(actionSetQuizCurrentQuestion(quizDatas.resume.lastQuestion));
      yield getCurrentOptions();
    } else {
      yield getNewQuestion({
        payload: {
          difficulty: quizDatas.resume.difficulty,
          category: category.payload.id,
        },
      });
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}

// CALLS
export function* onGetNewQuestion() {
  yield takeLatest(QuizActionsTypes.SAGA_GET_NEW_QUESTION, getNewQuestion);
}
export function* onCheckAnswer() {
  yield takeLatest(QuizActionsTypes.SAGA_CHECK_ANSWER, checkAnswer);
}
export function* onUpdateQuizReport() {
  yield takeLatest(QuizActionsTypes.SAGA_UPDATE_QUIZ_REPORT, updateQuizReport);
}
export function* onUpdateQuizResumeReport() {
  yield takeLatest(
    QuizActionsTypes.SAGA_UPDATE_QUIZ_RESUME_REPORT,
    updateQuizResumeReport
  );
}
export function* onGoToNextQuestion() {
  yield takeLatest(QuizActionsTypes.SAGA_GO_TO_NEXT_QUESTION, goToNextQuestion);
}
export function* onFinishQuiz() {
  yield takeLatest(QuizActionsTypes.SAGA_FINISH_QUIZ, finishQuiz);
}
export function* onResumeQuiz() {
  yield takeLatest(QuizActionsTypes.SAGA_RESUME_QUIZ, resumeQuiz);
}

export function* quizSagas() {
  yield all([
    call(onGetNewQuestion),
    call(onCheckAnswer),
    call(onUpdateQuizReport),
    call(onUpdateQuizResumeReport),
    call(onGoToNextQuestion),
    call(onFinishQuiz),
    call(onResumeQuiz),
  ]);
}
