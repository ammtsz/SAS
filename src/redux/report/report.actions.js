import { ReportActionsTypes } from "./report.types";

//REDUCERS
export const actionSetReportReview = (data) => ({
  type: ReportActionsTypes.REDUCER_SET_REPORT_REVIEW,
  payload: data,
});
export const actionSetReportCategory = (data) => ({
  type: ReportActionsTypes.REDUCER_SET_REPORT_CATEGORY,
  payload: data,
});
export const actionSetReportReviewCategory = (data) => ({
  type: ReportActionsTypes.REDUCER_SET_REPORT_REVIEW_CATEGORY,
  payload: data,
});
export const actionResetReport = () => ({
  type: ReportActionsTypes.REDUCER_RESET_REPORT,
});
