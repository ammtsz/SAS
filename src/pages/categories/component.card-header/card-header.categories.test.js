import React from "react";
import CardHeader from "./card-header.categories.component";

import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";

import { actionSetReportReviewCategory } from "../../../redux/report/report.actions";
import { actionResumeQuiz } from "../../../redux/quiz/quiz.actions";

describe("Component categories-page <CardHeader />", () => {
  const mockStore = configureMockStore();

  const mockHistory = {
    push: jest.fn(),
    location: { pathname: "/" },
  };

  describe("Category completed: > 0 and < 10 questions (quiz in progress)", () => {
    const mockReports = {
      1: {
        category_name: "",
        questions_datas: [],
        results: {
          difficulty: "",
          rights: 3,
          wrongs: 2,
        },
      },
    };
    const store = mockStore({
      categories: {
        reports:mockReports,
      },
      quiz: {},
    });
    store.dispatch = jest.fn();

    const mockCategory = { id: 1, name: "Books", completed: 5 };

    const mockProps = {
      category: mockCategory,
      history: mockHistory
    };

    let component;
    beforeEach(() => {
      component = render(
        <Provider store={store}>
          <MemoryRouter>
            <CardHeader {...mockProps} />
          </MemoryRouter>
        </Provider>
      );
    });
    
    afterEach(() => {
      cleanup()
    })
    
    it("should render CardHeader", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("should contain 'continue' text", () => {
      expect(screen.getByText("continue")).toBeInTheDocument();
    });

    it("should contain '<ProgressBar />' component", () => {
      expect(screen.getByTestId("progress-bar")).toBeInTheDocument();
    });

    it("should print 5 colored units in '<ProgressBar />' component", () => {
      expect(screen.getAllByTestId("progress-bar-colored").length).toBe(5);
    });

    it("should NOT contain results text '3/5'", () => {
      expect(screen.queryByTestId("card-header-results")).toEqual(null);
    });

    it("should fire 'resumeQuiz()' when click on 'continue' text", () => {
      fireEvent.click(screen.getByText("continue"));
      expect(store.dispatch).toHaveBeenCalledTimes(1);
      expect(store.dispatch).toHaveBeenCalledWith(actionResumeQuiz(mockCategory));
    });

    it("should redirect to '/quiz' page when click on 'continue' text", () => {
      fireEvent.click(screen.getByText("continue"));
      expect(mockProps.history.push).toHaveBeenCalledTimes(1)
      expect(mockProps.history.push).toHaveBeenCalledWith("/quiz")
    })
  });

  describe("Category completed: 10 questions (finished quiz)", () => {
    const reports = {
      1: {
        category_name: "",
        questions_datas: [],
        results: {
          difficulty: "",
          rights: 7,
          wrongs: 3,
        },
      },
    };
    const store = mockStore({
      categories: {
        reports,
      },
      quiz: {},
    });
    store.dispatch = jest.fn()

    const mockCategory = { id: 1, name: "Books", completed: 10 };
    
    const mockProps = {
      category: mockCategory,
      history: mockHistory,
    };

    let component
    beforeEach(() => {
      component = render(
        <Provider store={store}>
          <CardHeader {...mockProps} />
        </Provider>
      );
    });

    afterEach(() => {
      cleanup()
    })

    it("should render CardHeader", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("should contain 'see report' text", () => {
      expect(screen.getByText("see report")).toBeInTheDocument();
    });

    it("should NOT contain '<ProgressBar />' component", () => {
      expect(screen.queryByTestId("progress-bar")).toEqual(null);
    });

    it("should contain '7/10' text", () => {
      expect(screen.getByText("7/10")).toBeInTheDocument();
      expect(screen.getByTestId("card-header-results")).toBeInTheDocument();
    });

    it("should fire 'setReportReviewCategory()' when click on 'see report' text", () => {
      fireEvent.click(screen.getByText("see report"));
      expect(store.dispatch).toHaveBeenCalledTimes(1);
      expect(store.dispatch).toHaveBeenCalledWith(actionSetReportReviewCategory(mockCategory));
    });

    it("should redirect to '/report' page when click on 'see report' text", () => {
      fireEvent.click(screen.getByText("see report"));
      expect(mockProps.history.push).toHaveBeenCalledTimes(1)
      expect(mockProps.history.push).toHaveBeenCalledWith("/report")
    })
  });
});
