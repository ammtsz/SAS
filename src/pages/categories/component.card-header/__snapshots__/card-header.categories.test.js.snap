// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component categories-page <CardHeader /> Category completed: > 0 and < 10 questions (quiz in progress) should render CardHeader 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ hesUtF"
  >
    <div
      aria-label="5 questions answered"
      class="sc-gsTCUz eiooQx progress-bar"
      data-testid="progress-bar"
    >
      <span
        class="sc-dlfnbm dyfDju colored "
        data-testid="progress-bar-colored "
      />
      <span
        class="sc-dlfnbm dyfDju colored "
        data-testid="progress-bar-colored "
      />
      <span
        class="sc-dlfnbm dyfDju colored "
        data-testid="progress-bar-colored "
      />
      <span
        class="sc-dlfnbm dyfDju colored "
        data-testid="progress-bar-colored "
      />
      <span
        class="sc-dlfnbm dyfDju colored "
        data-testid="progress-bar-colored "
      />
      <span
        class="sc-dlfnbm dyfDju  "
        data-testid=" "
      />
      <span
        class="sc-dlfnbm dyfDju  "
        data-testid=" "
      />
      <span
        class="sc-dlfnbm dyfDju  "
        data-testid=" "
      />
      <span
        class="sc-dlfnbm dyfDju  "
        data-testid=" "
      />
      <span
        class="sc-dlfnbm dyfDju  "
        data-testid=" "
      />
    </div>
    <span
      aria-label="continue Books quiz (5 questions answered)"
      class="category-card__link"
      role="button"
      tabindex="0"
    >
      continue
    </span>
  </div>
</DocumentFragment>
`;

exports[`Component categories-page <CardHeader /> Category completed: 10 questions (finished quiz) should render CardHeader 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ hesUtF"
  >
    <div
      data-testid="card-header-results"
    >
      7/10
    </div>
    <span
      aria-label="see report for Books (quiz completed)"
      class="category-card__link"
      role="button"
      tabindex="0"
    >
      see report
    </span>
  </div>
</DocumentFragment>
`;
